ledger-service/http-json/src/main/scala/com/digitalasset/http/Endpoints.scala

```diff
@@ -403,25 +413,35 @@ class Endpoints(
   def query(req: HttpRequest)(implicit
       lc: LoggingContextOf[InstanceUUID with RequestID]
   ): Future[Error \/ SearchResult[Error \/ JsValue]] =
-    inputAndJwtPayload[JwtPayload](req).map {
-      _.flatMap { case (jwt, jwtPayload, reqBody) =>
-        withJwtPayloadLoggingContext(jwtPayload) { implicit lc =>
-          SprayJson
-            .decode[domain.GetActiveContractsRequest](reqBody)
-            .liftErr[Error](InvalidUserInput)
-            .map { cmd =>
-              val result: SearchResult[ContractsService.Error \/ domain.ActiveContract[JsValue]] =
-                contractsService.search(jwt, jwtPayload, cmd)
-
-              domain.SyncResponse.covariant.map(result) { source =>
-                source
-                  .via(handleSourceFailure)
-                  .map(_.flatMap(toJsValue[domain.ActiveContract[JsValue]](_)))
-              }
-            }
-        }
-      }
-    }
+    inputAndJwtPayload[JwtPayload](req)
+      .flatMap(it =>
+        either(it).flatMap { case (jwt, jwtPayload, reqBody) =>
+          withJwtPayloadLoggingContext(jwtPayload) { implicit lc =>
+            eitherT(
+              SprayJson
+                .decode[domain.GetActiveContractsRequest](reqBody)
+                .liftErr[Error](InvalidUserInput)
+                .map { cmd =>
+                  withEnrichedLoggingContext(
+                    LoggingContextOf.label[domain.GetActiveContractsRequest],
+                    "cmd" -> cmd.toString,
+                  ).run { implicit lc =>
+                    logger.debug(s"Processing a query request")
+                    contractsService
+                      .search(jwt, jwtPayload, cmd)
+                      .map(
+                        domain.SyncResponse.covariant.map(_)(
+                          _.via(handleSourceFailure)
+                            .map(_.flatMap(toJsValue[domain.ActiveContract[JsValue]](_)))
+                        )
+                      )
+                  }
+                }
+                .sequence
```

Comment:
generally I'd start looking to break things into more isolated/well contained methods when the nesting runs this deep as it becomes harder to understand the encapsulation of functions.

Reply:
yeah that function truly turned into a mess :smile: